Add public docs-search skill via Kapa MCP (#46)
AdonaiVera/fiftyone-skills: skills/fiftyone-sdk-guidance/SKILL.md

---
name: fiftyone-sdk-guidance

> description: Use when the user asks how to do something in the FiftyOne Python SDK, asks a docs question, or when no skill or operator is available to accomplish their goal, including requests like "how do I do X in Python", "show me the SDK for filtering", "there's no operator for this, can I do it in code", "what's the API for embeddings", "write me a Python script to do Y". Searches live FiftyOne documentation via an optional Kapa.ai-powered MCP connection and returns runnable code examples, falling back to training knowledge when that connection isn't available.

Reviewer: A couple thoughts/questions on the description (nothing blocks, just for your consideration):
* Do you want this to specifically trigger for when user "asks" about something? Seems like this could be a good opportunity to have a code agent fetch relevant documentation whenever needed, even if unprompted
* How much of the information in "optional Kapa.ai-powered MCP connection and returns runnable code examples, falling back to training knowledge when that connection isn't available" do you think is required for the code agent to trigger on this? It seems a little verbose to me.
* Expanding this to include issues situations where a code agent runs into unexpected or unknown API for FiftyOne might be helpful

Author: Usually should lean more toward a question, but good point, this could help, that's why I added  this in the first time "or when no skill or operator is available to accomplish their goal." But I added better wording in this commit, let me know what you think.

> optional Kapa.ai-powered MCP connection and returns runnable code examples, falling back to training knowledge when that connection isn't available.

I removed this, this is an instruction should not be here, good catch

---

# FiftyOne SDK Guidance

Answers FiftyOne SDK and docs questions by searching live documentation and returning accurate, runnable Python code. Covers two scenarios: direct SDK questions and fallback guidance when no operator or skill exists for the user's goal.

## Key Directives

**ALWAYS follow these rules, no exceptions:**

> ### 1. Always search docs first; never answer from training knowledge alone

Reviewer: Should this mention using the MCP for documentation searches specifically?

Author: Yes, we call that MCP the search docs, especifically "search_fifty_one_knowledge_sources", I think its better to keep it like that

```
search_fifty_one_knowledge_sources(query="How do I <user goal> in the FiftyOne Python SDK?")
```
FiftyOne's API changes frequently. Training knowledge may be outdated. Call `search_fifty_one_knowledge_sources` for every FiftyOne-specific question before answering.

**The query must be a complete natural-language sentence, not keywords.** This tool does semantic retrieval, not keyword search; its own schema requires "a single, well-formed natural-language query" that "must be a complete sentence." `query="filter confidence Python"` retrieves worse matches than `query="How do I filter detections by confidence threshold in the FiftyOne Python SDK?"`.

### 2. Return runnable Python code, not pseudocode, and cite your source

> All code examples must be complete enough to run. Include imports. Use real FiftyOne API patterns from the search results, not invented method names. Cite the `source_url` of the result(s) you drew from so the user can verify against live docs.

Reviewer: If there are no specific files/directories/datasets named, wonder if adding a "include constants for the user to fill in where values are unknown" to help avoid hunting down where the agent filled in random values.

Author: Oh in this case works exactly as the docs, they use a placeholder for dataset names or variables, like here:
<img width="802" height="859" alt="Screenshot 2026-07-09 at 11 23 25 AM" src="https://github.com/user-attachments/assets/a6bf9819-667e-4666-ba02-01961b58b3af" />

Reviewer: It's really just a nitpick and maybe trying to save on tokens (and avoid headaches if manual). I was just thinking that if a value is used multiple times in an example, it would be simpler to use a top-level constant. Something like this:

```python
import fiftyone as fo

MY_PATH = "path/to/dataset"  # <-- only have to change here

dataset = fo.Dataset.from_dir(MY_PATH, dataset_type=fo.types.COCODetectionDataset)
# ...
for sample in dataset.iter_samples():
    if not MY_PATH in sample.filepath:
        # ...

```

Like I mentioned, it's not as important, just wanted to distinguish what I originally intended to communicate.

### 3. For fallback scenarios, acknowledge the gap explicitly

> If the user was trying to use the App or operators and that path is unavailable, say so clearly:

Reviewer: ```suggestion
If the user is attempting to perform an unsupported or limited feature in the FiftyOne Application or operators, state clearly that this is the case and provide a supported route using the Python SDK where available:
```

```
"There's no built-in operator for this; here's how to do it with the Python SDK instead."
```

### 4. If one search isn't enough, search again with a refined query

> SDK questions often need 2–3 searches to find the right method, argument, or pattern. Search until you have a concrete answer.

Reviewer: ```suggestion
SDK questions may need at least 2–3 searches to find the right method, argument, or pattern. Search until you have a concrete answer or have concrete evidence that the feature is not available or included in the documentation.
```

### 5. Never make up API method names or field paths

> If the docs search doesn't return a clear answer, say so. Do not fabricate methods that look plausible.

Reviewer: ```suggestion
If the docs search doesn't return a clear answer, say so. Do not fabricate, assume, or guess methods that look plausible.
```

### 6. Degrade gracefully when the docs tool isn't connected

> This skill has no hard MCP dependency. If `search_fifty_one_knowledge_sources` isn't available in the current session, answer from training knowledge and say so; never block the user on a missing connection.

Reviewer: Perhaps directing to make calls to a web-search or web-fetch tool of `https://docs.voxel51.com` (if/when available) could be another alternative before a full fallback to the training data?

Author: Actually, your idea makes a lot of sense, since training knowledge is often wrong, and this works for the opensource skills where the agent can search the internet. I updated the three places where we mention training knowledge and changed it to web fetch retrieval.

### 7. Offer to self-install the docs connection, don't just tell the user to do it

> The first time this skill triggers and `search_fifty_one_knowledge_sources` isn't available, don't just leave the user to figure out setup themselves.

Reviewer: Should there be a zeroth directive?

> ### 0. Verify the 'fiftyone-docs' MCP server is available
> Use all tools available from the  'fiftyone-docs' MCP server as your primary resource for collecting information on FiftyOne. If the MCP server isn't available, see directive (7) about prompting for setup, otherwise follow directive (6) when user doesn't want the MCP server configured.

Author: Its fine

    }
  }
}
```

---

## Complete Workflow

### Step 1. Understand the user's goal

Identify:
- What they want to accomplish (e.g. filter, export, embed, annotate)
- Whether this is a pure SDK question or a fallback from a failed operator/plugin path
- What dataset or data type they're working with

### Step 2. Search the FiftyOne documentation

```
search_fifty_one_knowledge_sources(query="<specific goal, phrased as a full question>")
```

Search iteratively if the first result is partial. Examples of well-formed queries (complete sentences, not keyword fragments):
- "How do I filter detections by confidence threshold in the FiftyOne Python SDK?"
- "How do I compute CLIP embeddings in FiftyOne without using a plugin?"
- "How do I export a FiftyOne dataset to COCO format in Python?"
- "How do I match samples in a FiftyOne DatasetView by field value?"

The tool always returns a fixed number of results (`{"results": [{"source_url": "...", "content": "..."}, ...]}`), ranked by relevance, even when nothing in the docs actually matches, so don't assume every returned chunk is on-topic. Skim `content` for real relevance before using it.

### Step 3. Build the code response

> From the search results, construct a runnable example:

Reviewer: ```suggestion
From the search results, construct a runnable example. REFERENCE: an example constructed, after fetching all relevant FiftyOne documentation, for a user wanting a FiftyOne view for the dataset's ground truth labels that have a confidence value of more than 0.7:
```

```python
import fiftyone as fo
import fiftyone.zoo as foz

# Load or open your dataset
dataset = fo.load_dataset("my-dataset")

# Filter examples, built from search result
view = dataset.filter_labels(
    "ground_truth",
    F("confidence") > 0.7
)
print(view)
```

Always:
- Use `import fiftyone as fo` at the top
- Use real method names from the search results
- Show the user how to apply this to their specific dataset context

> - Cite the `source_url`(s) the code was drawn from, e.g. `**Source:** [Confidence thresholding in Python](https://docs.voxel51.com/...)`

Reviewer: ```suggestion
- Cite the `source_url`(s) the code was drawn from, e.g. `**Source:** [Confidence thresholding in Python](https://docs.voxel51.com/...)`; all URLs must be:
  * valid, HTTP 200 response
  * specific, link to the relevant section of the page, never point to top-level URLs and make the user search for where the information was found
  * relevant, URLs must be relevant to the topic/query, a valid and specific URL isn't helpful if it's not relevant to the user's goal/query
```

### Step 4. Explain the key methods

> After the code block, briefly explain the 1–2 most important methods:

Reviewer: ```suggestion
After the code block, briefly explain the most salient FiftyOne methods:
```

Reviewer: I fee like limiting 1-2 might end up dropping too much, also specifying it's for FiftyOne methods should help avoid explaining generic Python or third party libraries.

- Common variations (e.g. `filter_labels` vs `match`)

### Step 5. Offer follow-up

```
What would you like to do next?
  1. Run this code in a Python session or notebook
  2. Adapt the example to a different field or condition
  3. Chain this with another operation (e.g. export, tag, annotate)
  4. Look up a related SDK topic
```

---

## Available Tools

| Tool | Input | Output | Notes |
|------|-------|--------|-------|
| `search_fifty_one_knowledge_sources` | `query`, a single, well-formed natural-language question. **Must be a complete sentence**, not keywords. | `{"results": [{"source_url": "...", "content": "..."}, ...]}`, a fixed-size list of markdown chunks, ranked by relevance | Semantic retrieval over `docs.voxel51.com` and related sources, powered by Kapa.ai. Optional; see Directive 7 above for how to connect it. Always cite `source_url`; always returns top-k results even for off-topic queries, so verify relevance before using a chunk. |

---

## Common Use Cases

### "How do I filter by label confidence in Python?"

```
search_fifty_one_knowledge_sources(query="How do I filter detections by confidence threshold in the FiftyOne Python SDK?")
```

> 

Reviewer: Should the relevant page or an example of what's expected to be found be shown as an interim result? This is a bit similar to the example from above without this, so if you don't want to expand on this, maybe this one could be removed?

Author: Umm ye could be redudant, I'm going to keep the other example that you suggest to make it simpler

```python
import fiftyone as fo
from fiftyone import ViewField as F

dataset = fo.load_dataset("my-dataset")
view = dataset.filter_labels("ground_truth", F("confidence") > 0.7)
```

> 

Reviewer: Should an explanation go here? Same as above, without adding explanation (as the skill instructs), this ends up just looking like the example from Step 3 above and is a bit redundant.

Author: Umm ye could be redudant, I'm going to keep the other example that you suggest to make it simpler

> ### "There's no operator for exporting, how do I export from the SDK?"

Reviewer: There is an in-App label export operator tho, so maybe this isn't the best example? Maybe something like "I want to modify my model's weights directly using FiftyOne" since that's not at all supported?

Author: Umm yes could be redudant, I'm going to keep the other example that you suggest to make it simpler

```
search_fifty_one_knowledge_sources(query="How do I export a FiftyOne dataset to COCO format using the Python SDK?")
```

```python
import fiftyone as fo

dataset = fo.load_dataset("my-dataset")
dataset.export(
    export_dir="/path/to/export",
    dataset_type=fo.types.COCODetectionDataset,
    label_field="ground_truth",
)
```

> ### "How do I compute embeddings without a plugin?"

Reviewer: Would this be a good example for no in-App functionality so use the Python SDK instead?

```md
### "I want to generate embeddings for my dataset using the FiftyOne App."

search_fifty_one_knowledge_sources(query="How can I use the FiftyOne App to compute 
embeddings on my dataset?")
 
The computing embeddings using the FiftyOne App is a feature of FiftyOne Enterprise. OSS 
FiftyOne must use the Python SDK. Let me search how the correct SDK methods.
 
search_fifty_one_knowledge_sources(query="Show me how to use the FiftyOne Python SDK 
for computing embeddings.")

search_fifty_one_knowledge_sources(query="What's the difference between 
`compute_embeddings` and `compute_patch_embeddings`?")

Do you want to generate patch embeddings or image-level embeddings? If your dataset 
uses localized annotations like detection bounding boxes, poly-lines, or segmentation 
masks, you may want to use patch embeddings, if you want to check for nearly identical 
or similar images, you'll want to use image embeddings.

search_fifty_one_knowledge_sources(query="What are the required and optional arguments 
for the `compute_embeddings` method?")
```

Author: Nice! I'll add it

```
search_fifty_one_knowledge_sources(query="How do I compute embeddings on a FiftyOne dataset from the Python SDK without using a plugin?")
search_fifty_one_knowledge_sources(query="How does fiftyone.brain.compute_similarity work?")
```

```python
import fiftyone as fo
import fiftyone.brain as fob

dataset = fo.load_dataset("my-dataset")
fob.compute_similarity(dataset, model="clip-vit-base32", brain_key="clip_sim")
```

### "How do I split my dataset in Python?"

```
search_fifty_one_knowledge_sources(query="How do I split a FiftyOne dataset into train and test sets in Python?")
```

```python
import fiftyone as fo
import random

dataset = fo.load_dataset("my-dataset")
ids = dataset.values("id")
random.shuffle(ids)
split = int(len(ids) * 0.8)

train_view = dataset.select(ids[:split])
val_view = dataset.select(ids[split:])
```

---

## Fallback Scenarios

### No plugin or operator installed

> The user tried to use an operator and got "operator not found" or the plugin is not enabled.

Reviewer: These could be two very different situations. One could be a "guessed" or "presumed" correct method (even by another LLM/agent), the other could be an uninstalled plugin or necessary plugin. Wonder if these could be worth splitting apart?

Author: But for the user perspective, we dont know that, we received the same message operator not found

Reviewer: The distinction for this example I was trying to make is that an agent could trigger on this and spend a lot of time searching for something that might not exist, but it assumes (confidently) that it exists. My thought was to do specify that it would be best to determine if it's due to a missing dependency or disabled/uninstalled plugin _first_, then search if determined that's not the case. 

Maybe the section header "No plugin or operator installed" is sufficient to distinguish the two cases. I just wanted to call out that the "operator not found" message could have more than one cause that's not explicit in this example.

```
"That feature requires the @voxel51/panels plugin, which isn't enabled in your deployment.
Here's how to accomplish the same thing using the FiftyOne Python SDK:"
```

> Then search and return the SDK path.

Reviewer: ```suggestion
Then execute a search query in the FiftyOne documentation and return the SDK path.
```

### Operator ran but didn't produce expected results

> Search for the underlying SDK method the operator wraps, then show the user how to call it directly so they can debug or customize it.

Reviewer: Maybe it could be helpful to have the agent confirm its understanding of what the user is expecting first? That could change the search query if the initial response missed the intended goal.

### User wants more control than an operator provides

```
search_fifty_one_knowledge_sources(query="What's the low-level FiftyOne Python SDK API for <goal>?")
```

Show the low-level SDK call that gives them full control.

---

## Troubleshooting

**"`search_fifty_one_knowledge_sources` is not available"**

No Kapa.ai MCP server is connected in this session; this tool is optional, not a hard requirement. Apply Directive 7: offer to add it automatically (Claude Code), or hand the user the manual config snippet for their assistant (other assistants). Until it's connected, answer using training knowledge but note the limitation:
```
"I can't search the live docs right now, so I'll answer from training knowledge.
For the most accurate and up-to-date API reference, check docs.voxel51.com."

> ```

Reviewer: Same point as above. Maybe it would be good to redirect to a web-search tool against https://docs.voxel51.com first (if web-search tool is available), then fallback to training knowledge?

Author: Yes sir applied it
